File: android/src/main/java/com/thebluealliance/androidclient/config/ConfigModule.java
package com.thebluealliance.androidclient.config;

import android.content.SharedPreferences;

import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.thebluealliance.androidclient.TbaLogger;

import javax.annotation.Nullable;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;
import dagger.hilt.migration.DisableInstallInCheck;

@InstallIn(SingletonComponent.class)
@Module
public class ConfigModule {

    @Provides @Singleton @Nullable
    public FirebaseRemoteConfig provideFirebaseRemoteConfig() {
        try {
            return FirebaseRemoteConfig.getInstance();
        } catch (IllegalStateException ex) {
            /* When there is no google-secrets.json file found, the library throws an exception
             * here which causes insta-crashes for us. Silently recover here...
             */
            TbaLogger.e("Unable to find google-secrets.json, disabling remote config");
            return null;
        }
    }

    @Provides @Singleton
    public AppConfig provideAppConfig(@Nullable FirebaseRemoteConfig config, SharedPreferences prefs) {
        return new AppConfig(config, prefs);
    }
}

File: app/src/main/kotlin/com/thebluealliance/android/di/ConfigModule.kt
package com.thebluealliance.android.di

import android.content.Context
import android.content.SharedPreferences
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.preferencesDataStore
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "settings")

@Module
@InstallIn(SingletonComponent::class)
object ConfigModule {

    @Provides
    @Singleton
    fun provideSharedPreferences(@ApplicationContext context: Context): SharedPreferences =
        context.getSharedPreferences("tba_config", Context.MODE_PRIVATE)

    @Provides
    @Singleton
    fun provideFirebaseRemoteConfig(): FirebaseRemoteConfig =
        FirebaseRemoteConfig.getInstance()

    @Provides
    @Singleton
    fun provideDataStore(@ApplicationContext context: Context): DataStore<Preferences> =
        context.dataStore
}
